Azdevify chat-with-your-data-solution-accelerator.

zedy-wj opened the pull request:
This pr is all changes to convert chat-with-your-data-solution-accelerator application into an Azd template.

You can currently deploy this application using `azd` by following the steps below:

1. Execute the command: `azd init -t zedy-wj/chat-with-your-data-solution-accelerator-azdevify -b azdevify` 
2. Execute the command: `azd up`
3. Wait a moment for the resource deployment to complete, click the Website endpoint and you will see:
<img width="1256" alt="image" src="https://github.com/jongio/chat-with-your-data-solution-accelerator/assets/81678720/f317ce61-151f-4875-b166-571608209371">

@jongio - for notification.

zedy-wj committed:
azdevify this sample and add azd new feature in devcontainer

zedy-wj committed:
instead parameters.json file with .bicepparam

jongio commented on azure.yaml:
```diff
@@ -0,0 +1,26 @@
+          interactive: true
+          continueOnError: false
+
+  WebsiteName_admin:
```
name it `adminweb`

jongio commented on azure.yaml:
```diff
@@ -0,0 +1,26 @@
+name: chat-with-your-data-solution-accelerator
+
+services:
+  Website:
```
```suggestion
  web:
```

jongio replied:
When you change this it will also need to be changes every where else in code, bicep tags, etc.

jongio commented on infra/app/function.bicep:
```diff
@@ -0,0 +1,90 @@
+param storageAccountName string = ''
+param tags object = {}
+
+param appSettings array = []
```
This should have secure() decorator

jongio commented on infra/app/function.bicep:
```diff
@@ -0,0 +1,90 @@
+param applicationInsightsName string = ''
+param runtimeName string
+param runtimeVersion string
+param ClientKey string
```
Keep consistent casing for params.  All camelCase

jongio commented on infra/app/function.bicep:
```diff
@@ -0,0 +1,90 @@
+param ContentSafetyName string = ''
+
+
+module Function '../core/host/functions.bicep' = {
```
```suggestion
module function '../core/host/functions.bicep' = {
```

jongio commented on infra/app/function.bicep:
```diff
@@ -0,0 +1,90 @@
+    storageAccountName: storageAccountName
+    runtimeName: runtimeName
+    runtimeVersion: runtimeVersion
+    appSettings: union(toObject(appSettings, entry => entry.name, entry => entry.value), {
```
Do we need to do this toObject here?  I don't think we do that in other cases:

jongio commented on infra/app/function.bicep:
```diff
@@ -0,0 +1,90 @@
+  }
+}
+
+resource storage 'Microsoft.Storage/storageAccounts@2021-09-01' existing = {
```
Each one of these "exisiting" calls is a costly ARM deployment.  So we want to minimize them.  Instead of this, you should be passing the keys in via app settings from main.bicep.

zedy-wj replied:
If this is done, the current way I can think of is to pass these `keys` as `output` from the bicep file in the module into `main.bicep`, which doesn‘t seem very secure. Any better suggestions? Please take a look at my [latest commit](https://github.com/jongio/chat-with-your-data-solution-accelerator/pull/2/commits/c00f55a8f5f9c9b4f1aa6318fe493ef3a31cf995#diff-495e118ae4bb45392efbfa5a2f99260e3e2f72da47e6fdfe73d929d5f07c6068R64). Is this okay?

jongio commented on infra/app/websiteadmin.bicep:
```diff
@@ -0,0 +1,70 @@
+param appSettings array = []
+param serviceName string = 'WebsiteName_admin'
+
+module websiteadmin '../core/host/appservice.bicep' = {
```
camel

zedy-wj committed:
processing format

zedy-wj committed:
keep consistent casing for params

zedy-wj committed:
move appsettings

jongio commented on infra/app/adminweb.bicep:
```diff
@@ -0,0 +1,71 @@
+    applicationInsightsName: applicationInsightsName
+    appServicePlanId: appServicePlanId
+    appSettings: union(appSettings, {
+      // AZURE_BLOB_ACCOUNT_KEY: storage.listKeys().keys[0].value
```
Please remove any unnecessary commented out code.

jongio commented on infra/app/adminweb.bicep:
```diff
@@ -0,0 +1,71 @@
+//   name: applicationInsightsName
+// }
+
+resource openAI 'Microsoft.CognitiveServices/accounts@2023-05-01' existing = {
+  name: azureOpenAIName
+}
+
+resource search 'Microsoft.Search/searchServices@2021-04-01-preview' existing = {
+  name: azureCognitiveSearchName
+}
```
Can't we pass these values into the module instead of doing an existing lookup?

jongio commented on infra/app/function.bicep:
```diff
@@ -0,0 +1,91 @@
+// param applicationInsightsName string = ''
+param runtimeName string
+param runtimeVersion string
+param clientKey string
```
This needs to be secure

jongio commented on infra/app/resources.bicep:
```diff
@@ -0,0 +1,87 @@
+  }
+}
+
+output AZURE_FORM_RECOGNIZER_KEY string =  formRecognizer.listKeys().key1
+output AZURE_CONTENT_SAFETY_KEY string =  contentSafety.listKeys().key1
```
Outputs cannot contain keys

jongio commented on infra/app/storage.bicep:
```diff
@@ -0,0 +1,64 @@
+  }
+}
+
+resource storageAccountNameDefaultBlobContainer 'Microsoft.Storage/storageAccounts/blobServices/containers@2021-08-01' = {
```
Move all of these as child resources of the storageAccount.

Remove all the dependsOn

jongio commented on infra/app/storage.bicep:
```diff
@@ -0,0 +1,64 @@
+param storageAccountName string = 'Enabled'
+param location string = ''
+param blobContainerName string = 'Enabled'
```
Why defaulting these to 'Enabled'?

jongio commented on infra/app/storage.bicep:
```diff
@@ -0,0 +1,64 @@
+  dependsOn: []
+}
+
+output StorageAccountId string = storageAccount.id
+output StorageAccountName_default_doc_processing_name string = storageAccountNameDefaultDocProcessing.name
+output AZURE_BLOB_ACCOUNT_KEY string =  storageAccount.listKeys().keys[0].value
```
Be consistent on output casing

jongio replied:
Outputs can't contain keys

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,421 @@
+param environmentName string
+
+@description('provide a 2-13 character prefix for all resources.')
+param resourcePrefix string = environmentName
```
Instead of using resourcePrifix like this, use the method of generating the prefix like we do in azd templates.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,421 @@
+param azureOpenAIEmbeddingModelName string = 'text-embedding-ada-002'
+
+@description('Azure Cognitive Search Resource')
+param azureCognitiveSearch string = '${resourcePrefix}-search'
```
Use "Name" in the param name if it is the name, see the azd temaplates for reference.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,421 @@
+    name: azureCognitiveSearch
+    location: location
+    tags: {
+      deployment : 'chatwithyourdata-sa'
```
This should be a param, not hardcoded.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,421 @@
+}
+
+module resources './app/resources.bicep' = {
+  name: 'AllResources'
```
This isn't "AllREsources".  Suggest 'resource'

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,421 @@
+  }
+}
+
+module resources './app/resources.bicep' = {
```
Why do we need a "resources" file?  Can they not be in main.bicep?

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,421 @@
+    sku: {
+      name: hostingPlanSku
+    }
+    kind: 'linux'
```
The default is linux so you shouldn't need this.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,421 @@
+      AZURE_CONTENT_SAFETY_KEY: resources.outputs.AZURE_CONTENT_SAFETY_KEY
+    }
+  }
+  dependsOn:[
```
You shouldn't need dependsOn if you are referencing the modules by the reference, not the name.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,421 @@
+      AZURE_CONTENT_SAFETY_KEY: resources.outputs.AZURE_CONTENT_SAFETY_KEY
+    }
+  }
+  dependsOn:[
```
Go through and figure out how to remove all dependsOn.  They really shouldn't be needed if referencing resources correctly.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,421 @@
+  params: {
+    applicationInsightsName: applicationInsightsName
+    location: location
+    tags: {
+      'hidden-link:${resourceId('Microsoft.Web/sites', applicationInsightsName)}': 'Resource'
+    }
```
Why is this needed?  Should this be something we add to core in azd?

zedy-wj replied:
I think this is retained when exporting the Azure Resource Manager JSON template, it's an internal tag and we shouldn't delete it. And I've tried deleting it, getting some unexpected errors. As for whether it should be added into azd core, I think it is not necessary, just like todo templates do not need this tag to link web app resources.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,421 @@
+    azureCognitiveSearchName: azureCognitiveSearch
+    // formRecognizerName: formRecognizerName
+    // contentSafetyName: contentSafetyName
+    runtimeName:'python'
+    runtimeVersion:'3.11'
```
Isn't this already in function.bicep?

zedy-wj committed:
remove some invalid changes

zedy-wj committed:
fix format issues

zedy-wj committed:
Remove depandsOn

zedy-wj committed:
remove resources file

zedy-wj committed:
remov keys from output

zedy-wj committed:
remove existing reference in module

zedy-wj commented:
@jongio - Updated according to your comments, please re-review this pr. Thanks!

jongio commented:
> 1. Execute the command: `azd init -t https://github.com/zedy-wj/chat-with-your-data-solution-accelerator-azdevify.git -b azdevify`

You don't need the full URL, `azd init -t zedy-wj/chat-with-your-data-solution-accelerator-azdevify -b azdevify` works

jongio commented on README.md:
```diff
@@ -104,6 +104,44 @@ Out-of-the-box, you can upload the following file types:
     
     ![A screenshot of the chat app.](./media/chat-app.png)
 
+## Running the sample using Azd template
```
```suggestion
## Running the sample using the Azure Developer CLI (azd) template
```

jongio commented on README.md:
```diff
@@ -104,6 +104,44 @@ Out-of-the-box, you can upload the following file types:
+
+The Azure Developer CLI (`azd`) is a developer-centric command-line interface (CLI) tool for creating Azure applications.
+
+You need to install it before running and deploying with Azure Developer CLI.
```
```suggestion
You need to install it before running and deploying with the Azure Developer CLI.
```

jongio commented on README.md:
````diff
@@ -104,6 +104,44 @@ Out-of-the-box, you can upload the following file types:
+curl -fsSL https://aka.ms/install-azd.sh | bash
+```
+
+After logging in with the following command, you will be able to use the azd cli to quickly provision and deploy the application.
````
```suggestion
After logging in with the following command, you will be able to use azd cli to quickly provision and deploy the application.
```

jongio commented on README.md:
````diff
@@ -104,6 +104,44 @@ Out-of-the-box, you can upload the following file types:
+```
+azd init -t jongio/chat-with-your-data-solution-accelerator
+```
+According to the prompt, enter an `env name`.
````
```suggestion
According to the prompt, enter an environment name.
```

jongio commented on README.md:
````diff
@@ -104,6 +104,44 @@ Out-of-the-box, you can upload the following file types:
+azd up 
+```
+
+According to the prompt, select `subscription` and `location`, these are the necessary parameters when you create resources. After that, choose a resource group created or create a new resource group. Wait a moment for the resource deployment to complete, click the Website endpoint and you will see the web app page.
````
```suggestion
According to the prompt, select `subscription` and `location`, these are the necessary parameters when you create resources. After that, choose a resource group or create a new resource group. Wait a moment for the resource deployment to complete, click the Website endpoint and you will see the web app page.
```

jongio commented on README.md:
```diff
@@ -104,6 +104,44 @@ Out-of-the-box, you can upload the following file types:
+
+You can also run the sample directly locally (See below).
 
 ## Development and run the accelerator locally
```
```suggestion
## Develop and run the accelerator locally
```

jongio commented on azure.yaml:
```diff
@@ -0,0 +1,26 @@
+# yaml-language-server: $schema=https://raw.githubusercontent.com/Azure/azure-dev/main/schemas/v1.0/azure.yaml.json
+
+name: chat-with-your-data-solution-accelerator
+
```
Include the metadata attributes like other projects

jongio commented on infra/app/adminweb.bicep:
```diff
@@ -0,0 +1,40 @@
+param appCommandLine string = 'python -m streamlit run Admin.py --server.port 8000 --server.address 0.0.0.0 --server.enableXsrfProtection false'
+param applicationInsightsName string = ''
+param keyVaultName string = ''
+param azureOpenAIName string = ''
```
```suggestion
param azureOpenAiName string = ''
```

jongio replied:
Update all openAI to openAi and all OpenAI to OpenAi

jongio commented on infra/app/adminweb.bicep:
```diff
@@ -0,0 +1,40 @@
+    applicationInsightsName: applicationInsightsName
+    appServicePlanId: appServicePlanId
+    appSettings: union(appSettings, {
+      AZURE_OPENAI_KEY: listKeys('Microsoft.CognitiveServices/accounts/${azureOpenAIName}', '2023-05-01').key1
+      AZURE_SEARCH_KEY: listAdminKeys('Microsoft.Search/searchServices/${azureCognitiveSearchName}', '2021-04-01-preview').primaryKey
```
This only works if the openai and search are in the same resource group.  You will have to pass the keys into this from main.bicep or better yet, put the keys in keyvault and then just store the names here to be retrieved by the application.

zedy-wj replied:
Hi, @jongio - We've tried a lot of things here on how to set `keys` to `appSettings`. Before making this change, I'd like to discuss best practices with you.

1.If we want to set `resources.keys` to `appSettings` in the `main.bicep` file, there are three methods we have tried so far:
- Create relevant resources directly in the main.bicep file, and then use `resources.listKeys().key1`. In this way, we cannot use templates in `core` to create related resources.
- Use `module` in `main.bicep` to reference related templates to create resources, but at this time, `resource.listKeys().key1` cannot be used because the resource is not created at the level of `main.bicep`; `listKeys(resourceId).key1` cannot be used either ( At present, the priority of the `listKeys` method is higher than that of module, unless we create the resource first).
- After we use the `module` to create the resource, use `existing` in `main.bicep` to call the relevant resources and then use `resource.listKeys().key1`. But this will also cause problems. The priority of `existing` is higher than `module`. We need to ensure that the resource is created in advance before we use `existing`.

Based on the results of our investigation so far, if you have any questions or if you have other best practices on how to set keys to appSettings in main.bicep, please let us know. Thank you very much!

2.About put the `keys` in `keyvault` and then just store the names here to be retrieved by the application:

- We also tried this in another azdevify [PR](https://github.com/Azure-Samples/azure-sql-db-session-recommender/pull/7/files/79acbab1a3fafca49c517befecc6188da1afb2c0#diff-ec8f897aa232a01946f69fd44fea3f1b350a0ae5fa1228d62bcb17bba705a239R45-R51). We refer to the approach in [todo template](https://github.com/Azure-Samples/todo-csharp-sql/blob/main/infra/core/database/cosmos/cosmos-account.bicep).  If we want to store the key of a resource, we need to reference `KeyVault` when creating the resource. Currently, the template for creating `openAi` in core does not have this function, which means that we cannot use the template in core when creating related resources.


Do you have any ideas on this? Our current idea is to add the function of storing `resource's keys` in the relevant templates in core: add a param `keyVaultName`, which is empty by default. If this value is empty, the resource `'Microsoft.KeyVault/vaults/secrets@2022-07-01'` will not be called.  If `keyVaultName` is passed in externally, the key of this resource can be stored in `KeyVault`. Do you think so?

v-xuto replied:
@jongio Any ideas about this issue?

jongio replied:
1. Add a parameter to the main.bicep that allows the user to either use RBAC or KEYS.  Call it AUTH_TYPE and default it to rbac.  And set @allowed to 'rbac' or 'keys'
2. If the auth type is rbac, then use roles.  If auth type is keys, then use keys.  

You can look up the keys with this listKeys method.  Example for storage.

```
listKeys(resourceId(subscription().subscriptionId, rgName, 'Microsoft.Storage/storageAccounts', storageAccountName), '2021-09-01').keys[0].value
```


It's like what you had before, but you also include the subscriptionId and RgName

zedy-wj replied:
Currently we use `KeyVault` to store all keys and then obtain the keys from the application. What do you think of this? If you insist on using `rbac`, the way of creating various `service clients` in the application code needs to be changed accordingly. Then the gap between local deployment and deployment using `azd` will be slightly too large.

jongio replied:
I see what you mean.  Could you come up with a version of the code that shows what it would look like to instantiate the objects based on rbac or keys?  Just so I can see what the code would look like.  Doesn't have to be for them all.

zedy-wj replied:
@jongio - Sorry for the delay. We have done some exploration and attempts on how to use `rbac`. Details as following:

Currently, a total of five `Keys` are used in our appSettings, which are: `AZURE_OPENAI_KEY`, `AZURE_SEARCH_KEY`, `AZURE_BLOB_ACCOUNT_KEY`, `AZURE_FORM_RECOGNIZER_KEY`, `AZURE_CONTENT_SAFETY_KEY`.

We divide it into three categories:

1. Cognitive Services (`AZURE_OPENAI_KEY`, `AZURE_FORM_RECOGNIZER_KEY`, `AZURE_CONTENT_SAFETY_KEY`):

- For this type of `Keys`, we can add rbac permissions to `openai`, and then use the following method to assign values to the `Keys`:
![image](https://github.com/jongio/chat-with-your-data-solution-accelerator/assets/81678720/67c1737f-7805-485a-ad73-e276a0924334)

2. AZURE_BLOB_ACCOUNT_KEY:

- We only need to add relevant permissions to `storage` and change the way to create `BlobServiceClient`:
![image](https://github.com/jongio/chat-with-your-data-solution-accelerator/assets/81678720/31b17301-1271-4f83-99ee-fc33cdf027f4)

3. AZURE_SEARCH_KEY:

- It needs to be used to create the request body in the [application](https://github.com/jongio/chat-with-your-data-solution-accelerator/blob/main/app.py#L79). We have not found a suitable replacement method yet: 
![image](https://github.com/jongio/chat-with-your-data-solution-accelerator/assets/81678720/a194f10b-3b65-4664-bed6-e98d825a72d0)If we want to change it, we can refer to [azure-search-openai-demo](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/app/backend/app.py), but the change will be large.

jongio replied:
Oh, I didn't know they aren't using the search SDK.  I'm asking them about that and will get back to you.

jongio commented on infra/app/function.bicep:
```diff
@@ -0,0 +1,61 @@
+    runtimeName: runtimeName
+    runtimeVersion: runtimeVersion
+    appSettings: union(appSettings, {
+      AZURE_OPENAI_KEY: listKeys('Microsoft.CognitiveServices/accounts/${azureOpenAIName}', '2023-05-01').key1
+      AZURE_SEARCH_KEY: listAdminKeys('Microsoft.Search/searchServices/${azureCognitiveSearchName}', '2021-04-01-preview').primaryKey
```
Same comment as above

jongio commented on infra/app/web.bicep:
```diff
@@ -0,0 +1,41 @@
+    appCommandLine: appCommandLine
+    applicationInsightsName: applicationInsightsName
+    appServicePlanId: appServicePlanId
+    appSettings: union(appSettings, {
+      AZURE_OPENAI_KEY: listKeys('Microsoft.CognitiveServices/accounts/${azureOpenAIName}', '2023-05-01').key1
+      AZURE_SEARCH_KEY: listAdminKeys('Microsoft.Search/searchServices/${azureCognitiveSearchName}', '2021-04-01-preview').primaryKey
```
Same comment as above

jongio commented on infra/core/ai/cognitiveservices.bicep:
```diff
@@ -0,0 +1,42 @@
+metadata description = 'Creates an Azure Cognitive Services instance.'
```
Is infra/core the same as azure-dev/infra/core?  Any changes needed for this app?

zedy-wj replied:
So far they are consistent and we haven't changed anything. There may be changes later, please refer to my [comment](https://github.com/jongio/chat-with-your-data-solution-accelerator/pull/2#discussion_r1392265065) here.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,483 @@
+param azureSearchUrlColumn string = 'url'
+
+@description('Name of Azure OpenAI Resource')
+param azureOpenAIResourceName string = '${environmentName}-openai-${resourceToken}'
```
If the original project uses OpenAI, then use that across the board.  If not, then use OpenAi, just be consistent

zedy-wj replied:
Okay, the original project uses OpenAI.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,483 @@
+    metadata: {}
+  }
+  dependsOn: []
+}
```
main.bicep should have all the outputs needed to run the application.  See other azd templates for examples

zedy-wj committed:
update ReadMe

jongio commented:
> 1. Execute the command: `azd init -t zedy-wj/chat-with-your-data-solution-accelerator-azdevify.git -b azdevify`

FYI - You don't need ".git" in the name

zedy-wj committed:
add rgName in listkeys

zedy-wj committed:
add use keyvault option

add keyvault option to get secret keys

remove invalid changes

add param useKeyVault

zedy-wj commented on infra/app/function.bicep:
```diff
@@ -0,0 +1,61 @@
+  }
+}
+
+resource functionNameDefaultClientKey 'Microsoft.Web/sites/host/functionKeys@2018-11-01' = {
+  name: '${name}/default/clientKey'
+  properties: {
+    name: 'ClientKey'
+    value: clientKey
+  }
+  dependsOn: [
+    function
+    waitFunctionDeploymentSection
+  ]
+}
```
It is a param required during [embed request](https://github.com/zedy-wj/chat-with-your-data-solution-accelerator-azdevify/blob/azdevify/backend/pages/01_Ingest_Data.py#L33).

zedy-wj committed:
set default value in .bicepparam

zedy-wj committed:
fix some bugs

jongio commented:
@zedy-wj  Please LMK when ready for re-review.  Thanks

zedy-wj commented:
@jongio If using `KeyVault or KEYS`, the code is now ready for re-review. If want to use `RBAC or KEYS`, there are currently some difficulties in obtaining the `search key`. For details, please refer to the comment [here](https://github.com/jongio/chat-with-your-data-solution-accelerator/pull/2#discussion_r1410444411). Do you think we need to make more efforts for `RBAC`?

jongio commented:
@zedy-wj - Please move forward to using the search SDK instead of REST calls.

jongio commented:
@zedy-wj - Please see this PR for example of using AZURE_AUTH_TYPE. Just as a reference https://github.com/microsoft/sample-app-aoai-chatGPT/pull/427

zedy-wj committed:
update rbac

ross-p-smith commented:
Hi @zedy-wj and @jongio - Is this PR going to make it to the main repo as I think this would be amazing to aid developers picking that accelerator up with the minimum of fuss. Some of the changes we are making as a team are going to start to conflict with this, as there are tasks to azdevify the repo. Is this PR in a state that we can take it and fix things up? I just worry that it's 100 commits behind main at this point. What are your thoughts?

zedy-wj commented:
Hi @ross-p-smith - Jongio is on vacation recently, and he should give a definite reply when he comes back. What we know so far is that this PR will be merged under Jongio's repo, and there is a high probability that it will eventually be merged into the main repo, but we are not sure.

zedy-wj committed:
update rbac

zedy-wj commented:
@jongio - The update of `rbac` has been completed, please re-review it.

> @zedy-wj - Please see this PR for example of using AZURE_AUTH_TYPE. Just as a reference https://github.com/microsoft/sample-app-aoai-chatGPT/pull/427

Regarding the part you mentioned before, we have not modified it. The reason is that after our investigation, the code related to `REST calls` is not used now. Currently, only the `/api/conversation/custom` interface is used in the `app.py` file. Therefore, you can simply understand that the search-related code is not used.

jongio commented on README.md:
```diff
@@ -104,8 +104,46 @@ Out-of-the-box, you can upload the following file types:
     
     ![A screenshot of the chat app.](./media/chat-app.png)
 
+## Running the sample using the Azure Developer CLI (azd) template
```
```suggestion
## Running the sample using the Azure Developer CLI (azd)
```

jongio commented on README.md:
````diff
@@ -104,8 +104,46 @@ Out-of-the-box, you can upload the following file types:
+
+Then, execute the `azd init` command to initialize the environment.
+```
+azd init -t jongio/chat-with-your-data-solution-accelerator
````
Leaving a comment here to remove `jongio` from this command

jongio commented on app.py:
```diff
@@ -23,10 +24,17 @@
 def static_file(path):
     return app.send_static_file(path)
 
+USE_RBAC = False
+
+if os.environ.get("AUTH_TYPE") == 'rbac':
+    USE_RBAC = True
```
@pamelafox to comment on use of global here.

pamelafox replied:
This code still uses an architecture that we try to avoid, where it stores all this in globals. It makes it hard to test and overall more brittle. The preferred approach is to use an app factory pattern. Here's an example PR that ported the other RAG app to an app factory:
https://github.com/Azure-Samples/azure-search-openai-demo/pull/495

That is a larger change so probably shouldnt be a part of this PR

zedy-wj replied:
We removed the global variable definition of `USE_RBAC` and instead obtained `AUTH_TYPE` from the environment variable.

jongio commented on backend/pages/01_Ingest_Data.py:
```diff
@@ -65,18 +79,28 @@ def upload_file(bytes_data: bytes, file_name: str, content_type: Optional[str] =
-    st.session_state['file_url'] = blob_client.url + '?' + generate_blob_sas(account_name, container_name, file_name,account_key=account_key,  permission="r", expiry=datetime.utcnow() + timedelta(hours=3))
+    if os.environ.get("AUTH_TYPE") == 'rbac':
+        credential = DefaultAzureCredential()
+        account_url = 'https://' + account_name +'.blob.core.windows.net/'
```
Can we use a format string here? f'http://{account_name}.blob.core.windows.net/'

zedy-wj replied:
Updated.

jongio commented on backend/utilities/helpers/AzureBlobStorageHelper.py:
```diff
@@ -1,27 +1,55 @@
+            self.account_name = account_name if account_name else env_helper.AZURE_BLOB_ACCOUNT_NAME
+            self.container_name : str = container_name if container_name else env_helper.AZURE_BLOB_CONTAINER_NAME
+            credential = DefaultAzureCredential()
+            account_url = 'https://' + self.account_name +'.blob.core.windows.net/'
```
f string comment here too

jongio commented on backend/utilities/helpers/AzureBlobStorageHelper.py:
```diff
@@ -80,8 +114,14 @@ def upsert_blob_metadata(self, file_name, metadata):
     def get_container_sas(self):
         # Generate a SAS URL to the container and return it
-        return "?" + generate_container_sas(account_name= self.account_name, container_name= self.container_name,account_key=self.account_key,  permission="r", expiry=datetime.utcnow() + timedelta(hours=1))
+        if self.use_rbac:
+            return "?" + generate_container_sas(account_name= self.account_name, container_name= self.container_name, user_delegation_key=self.user_delegation_key,  permission="r", expiry=datetime.utcnow() + timedelta(hours=1))
+        else:
+            return "?" + generate_container_sas(account_name= self.account_name, container_name= self.container_name,account_key=self.account_key,  permission="r", expiry=datetime.utcnow() + timedelta(hours=1))
```
I feel like we need to abstract this "auth type" logic into a class and then have the class manage what is returned.

I'm seeing too much duplicated forking code.  Can you come up with a better design?

zedy-wj replied:
We have updated the code and no longer repeatedly determine whether to use `rbac`.

jongio commented on infra/app/adminweb.bicep:
```diff
@@ -0,0 +1,47 @@
+    appServicePlanId: appServicePlanId
+    appSettings: union(appSettings, {
+      AUTH_TYPE: authType
+      AZURE_OPENAI_KEY: listKeys(resourceId(subscription().subscriptionId, rgName, 'Microsoft.CognitiveServices/accounts', azureOpenAIName), '2023-05-01').key1
```
Can you test something?  Can you try to get the resourceGroupName with 'resourceGroup().name"?  

I think the module is scoped to a resourceGroup which is set in the main.bicep file.  So, if the above works, then we may not have to pass in the rgName.

zedy-wj replied:
Sounds great! We tried this and it worked fine. Updated.

jongio commented on infra/app/storage.bicep:
```diff
@@ -0,0 +1,56 @@
+param location string
+param blobContainerName string
+
+resource storageAccount 'Microsoft.Storage/storageAccounts@2021-09-01' = {
```
You should be able to use infra/core for this.

zedy-wj replied:
In the `storage.bicep` file, not only `StorageAccount` and the corresponding `BlobServices` are created, but also the `QueueServices` with `StorageAccount` as the parent is created. infra/core does not support this part.

jongio replied:
Okay, in this case, please update the /infra/core modules to support this in this project, and then we can do a PR into azure-dev repo later.

zedy-wj replied:
Updated.

jongio commented on infra/core/security/keyvault-access.bicep:
```diff
@@ -0,0 +1,22 @@
+metadata description = 'Assigns an Azure Key Vault access policy.'
```
Please diff infra/core with latest in azure-dev repo, just to make sure we haven't diverged.

zedy-wj replied:
Consistent.

jongio commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,31 @@
+{
```
Can you investigate this flow:

1. Use RBAC
2. Store and Use Keys in KeyVault.  If the KV isn't specified, then try to read the values from the service with listKeys.

The above adds more complexity, but allows the user to store the keys in KV as an option.

ross-p-smith replied:
Just linking this as we should revisit the Speech Service using this method too: - https://github.com/Azure-Samples/chat-with-your-data-solution-accelerator/issues/101 (not for this PR!)

zedy-wj replied:
Use of `KeyVault` has been added and is currently testing well.

jongio replied:
Sorry for the back and forth on this.  Let's do this:

Default to using rbac.

If AUTH_TYPE is rbac, then assume the developer wants to also use key vault.  Put the keys in keyvault and then read them from keyvault.

If the user sets AUTH_TYPE to keys, then read the keys from the environment and set the keys using listKeys function.

I think that will be cleaner.

zedy-wj replied:
In our understanding, if `AUTH_TYPE` is `rbac`, then some `App Services` and `User` should be assigned corresponding permissions in the `main.bicep` file. This means that keys will no longer be needed in the code, but the service will be accessed directly through something like `DefaultAzureCredential()`.

So regarding what you mention here, `If AUTH_TYPE is rbac, then assume the developer wants to also use key vault.`, which confuses us. Because in our concept, one of them is enough.

If we have any misunderstandings about `rbac` or your comments, please let me know, thanks.

jongio replied:
Yes, only use a key vault if there's no RBAC option for the service.

pamelafox commented on app.py:
```diff
@@ -23,10 +24,17 @@
+if os.environ.get("AUTH_TYPE") == 'rbac':
+    USE_RBAC = True
+    credential = DefaultAzureCredential()
+    openai_token = credential.get_token("https://cognitiveservices.azure.com/.default")
```
Looks like this also uses old openAI SDK with tokens that must be manually refreshed. Recommend moving to new SDK to take care of token refresh for you.

zedy-wj replied:
Upgrade `openai` to V1.

pamelafox commented on app.py:
```diff
@@ -39,7 +47,7 @@ def static_file(path):
 AZURE_OPENAI_RESOURCE = os.environ.get("AZURE_OPENAI_RESOURCE")
 AZURE_OPENAI_MODEL = os.environ.get("AZURE_OPENAI_MODEL")
-AZURE_OPENAI_KEY = os.environ.get("AZURE_OPENAI_KEY")
+AZURE_OPENAI_KEY = openai_token.token if USE_RBAC else os.environ.get("AZURE_OPENAI_KEY")
```
This token will need to manually refreshed every so often if using old SDK, see bottom ofhttps://blog.pamelafox.org/2023/09/best-practices-for-openai-chat-apps-go.html

zedy-wj replied:
Upgrade `openai` to V1.

pamelafox commented on azure.yaml:
```diff
@@ -0,0 +1,27 @@
+
+name: chat-with-your-data-solution-accelerator
+metadata:
+  template: chat-with-your-data-solution-accelerator
```
template IDs should have version numbers, right?

pamelafox commented on backend/utilities/helpers/AzureBlobStorageHelper.py:
```diff
@@ -1,27 +1,55 @@
-        self.blob_service_client : BlobServiceClient = BlobServiceClient.from_connection_string(self.connect_str)
-
-
+        self.use_rbac = False
```
Should this logic live in EnvHelper?

zedy-wj replied:
We introduced an `AUTH_TYPE` in `EnvHelper` to facilitate use in other files.

pamelafox commented on backend/utilities/helpers/AzureBlobStorageHelper.py:
```diff
@@ -1,27 +1,55 @@
-from azure.storage.blob import BlobServiceClient, generate_blob_sas, generate_container_sas, ContentSettings
+from azure.storage.blob import BlobServiceClient, generate_blob_sas, generate_container_sas, ContentSettings, UserDelegationKey
 from .EnvHelper import EnvHelper
+from datetime import datetime, timedelta
```
I know this repo doesn't enforce import sorting but it's always nice to do it anyway- stdlib imports would be at top of file, then third party, then local.

zedy-wj committed:
fix some format and update infra core

ChenxiJiang333 committed:
Update openai to v1

zedy-wj committed:
use resourcegroup.name() instead pass in rhName

zedy-wj committed:
Merge pull request #1 from zedy-wj/openaiv1

Update openai to v1

zedy-wj committed:
Revert "Update openai to v1"

zedy-wj committed:
Merge pull request #2 from zedy-wj/revert-1-openaiv1

Revert "Update openai to v1"

ChenxiJiang333 committed:
add token provider

ChenxiJiang333 committed:
Merge branch 'azdevify' into openaiv1

zedy-wj committed:
Merge pull request #3 from zedy-wj/openaiv1

add token provider

ChenxiJiang333 committed:
Update requirements.txt

zedy-wj committed:
Merge pull request #4 from zedy-wj/ChenxiJiang333-patch-1

Update requirements.txt

zedy-wj committed:
revert

ChenxiJiang333 committed:
fix chat

ChenxiJiang333 committed:
Update TextProcessingTool.py

zedy-wj committed:
Merge pull request #5 from zedy-wj/newcommit

Upgrade openai to V1

jongio commented:
@zedy-wj - Please LMK when you are ready for re-review.

zedy-wj committed:
update KeyVault options

zedy-wj committed:
abstract auth type with define method

zedy-wj committed:
remove invalid infra/core

zedy-wj commented:
@jongio - We have updated all the code based on your comments, and updated infra/core to be consistent with azure-dev, please re-review it, thanks!

jongio commented on README.md:
````diff
@@ -104,8 +104,46 @@ Out-of-the-box, you can upload the following file types:
+azd auth login
+```
+
+Then, execute the `azd init` command to initialize the environment.
````
Add a comment here.  You do not need to run this command if you already have the code or have opened this in a Codespace or DevContainer.

jongio commented on README.md:
````diff
@@ -104,8 +104,46 @@ Out-of-the-box, you can upload the following file types:
+```
+azd init -t chat-with-your-data-solution-accelerator
+```
+According to the prompt, enter an environment name.
````
```suggestion
Enter an environment name.
```

jongio commented on README.md:
````diff
@@ -104,8 +104,46 @@ Out-of-the-box, you can upload the following file types:
+azd up 
+```
+
+According to the prompt, select `subscription` and `location`, these are the necessary parameters when you create resources. After that, choose a resource group or create a new resource group. Wait a moment for the resource deployment to complete, click the Website endpoint and you will see the web app page.
````
```suggestion
Select your desired `subscription` and `location`. Then choose a resource group or create a new resource group. Wait a moment for the resource deployment to complete, click the Website endpoint and you will see the web app page.
```

jongio commented on infra/app/function.bicep:
```diff
@@ -0,0 +1,78 @@
+param formRecognizerName string = ''
+param contentSafetyName string = ''
+param useKeyVault bool
+param openAIKey string = ''
+param storageAccountKey string = ''
+param formRecognizerKey string = ''
+param searchKey string = ''
+param contentSafetyKey string = ''
+param keyVaultEndpoint string = ''
```
Any parameter that is a key needs to be marked with "secure()" decorator.

zedy-wj replied:
The keys here are the key names stored in the `KeyVault`, not the actual key values.

jongio commented on infra/app/storekeys.bicep:
```diff
@@ -0,0 +1,63 @@
+  name: keyVaultName
+}
+
+output CONTENT_SAFETY_KEY string = contentSafetyKeySecret.name
+output FORM_RECOGNIZER_KEY string = formRecognizerKeySecret.name
+output SEARCH_KEY string = searchKeySecret.name
+output OPENAI_KEY string = openAIKeySecret.name
+output STORAGE_ACCOUNT_KEY string = storageAccountKeySecret.name
```
You cannot export keys from a bicep module.  You'll have to get them when they are needed in the module they are needed in.

zedy-wj replied:
The keys here are the key names stored in the `KeyVault`, not the actual key values.

jongio commented on infra/main.bicep:
```diff
@@ -0,0 +1,703 @@
+
+output APPLICATIONINSIGHTS_CONNECTION_STRING string = monitoring.outputs.applicationInsightsConnectionString
+output APPINSIGHTS_INSTRUMENTATIONKEY string = monitoring.outputs.applicationInsightsInstrumentationKey
+output AZURE_KEY_VAULT_ENDPOINT string = useKeyVault ? keyvault.outputs.endpoint : ''
+output AZURE_KEY_VAULT_NAME string = useKeyVault || authType == 'rbac' ? keyvault.outputs.name : ''
+output AZURE_LOCATION string = location
+output AZURE_TENANT_ID string = tenant().tenantId
+output AZURE_CONTENT_SAFETY_ENDPOINT string = contentsafety.outputs.endpoint
+output AZURE_SEARCH_SERVICE string = search.outputs.endpoint
+output AZURE_FORM_RECOGNIZER_ENDPOINT string = formrecognizer.outputs.endpoint
+output AZURE_SEARCH_USE_SEMANTIC_SEARCH string = azureSearchUseSemanticSearch
+output AZURE_SEARCH_SEMANTIC_SEARCH_CONFIG string = azureSearchSemanticSearchConfig
+output AZURE_SEARCH_INDEX_IS_PRECHUNKED string = azureSearchIndexIsPrechunked
+output AZURE_SEARCH_TOP_K string = azureSearchTopK
+output AZURE_SEARCH_ENABLE_IN_DOMAIN string = azureSearchEnableInDomain
+output AZURE_SEARCH_CONTENT_COLUMNS string = azureSearchContentColumns
+output AZURE_SEARCH_FILENAME_COLUMN string = azureSearchFilenameColumn
+output AZURE_SEARCH_TITLE_COLUMN string = azureSearchTitleColumn
+output AZURE_SEARCH_URL_COLUMN string = azureSearchUrlColumn
+output AZURE_OPENAI_MODEL_NAME string = azureOpenAIModelName
+output AZURE_OPENAI_STREAM string = azureOpenAIStream
+output AZURE_OPENAI_SYSTEM_MESSAGE string = azureOpenAISystemMessage
+output AZURE_OPENAI_STOP_SEQUENCE string = azureOpenAIStopSequence
+output AZURE_OPENAI_MAX_TOKENS string = azureOpenAIMaxTokens
+output AZURE_OPENAI_TOP_P string = azureOpenAITopP
+output AZURE_OPENAI_TEMPERATURE string = azureOpenAITemperature
+output AZURE_SEARCH_INDEX string = azureSearchIndex
+output AZURE_OPENAI_API_VERSION string = azureOpenAIApiVersion
+output DOCUMENT_PROCESSING_QUEUE_NAME string = queueName
+output AZURE_BLOB_CONTAINER_NAME string = blobContainerName
+output AZURE_BLOB_ACCOUNT_NAME string = storageAccountName
+output AZURE_OPENAI_RESOURCE string = azureOpenAIResourceName
+output AZURE_OPENAI_EMBEDDING_MODEL string = azureOpenAIEmbeddingModel
+output AZURE_OPENAI_MODEL string = azureOpenAIModel
+output USE_KEY_VAULT bool = useKeyVault
+output AZURE_OPENAI_KEY string = useKeyVault ? storekeys.outputs.OPENAI_KEY : ''
+output AZURE_BLOB_ACCOUNT_KEY string = useKeyVault ? storekeys.outputs.STORAGE_ACCOUNT_KEY: ''
+output AZURE_FORM_RECOGNIZER_KEY string = useKeyVault ? storekeys.outputs.FORM_RECOGNIZER_KEY: ''
+output AZURE_SEARCH_KEY string = useKeyVault ? storekeys.outputs.SEARCH_KEY: ''
+output AZURE_CONTENT_SAFETY_KEY string = useKeyVault ? storekeys.outputs.CONTENT_SAFETY_KEY: ''
```
You cannot output keys in bicep

zedy-wj replied:
The keys here are the key names stored in the `KeyVault`, not the actual key values.

jongio replied:
Please update the output to *KEY_NAME

zedy-wj committed:
update readme and storage.bicep in infra/core

zedy-wj commented:
@jongio - Updated according to your comments. For some things that haven’t been changed yet, please see my reply to your comments above, thanks!

zedy-wj committed:
Update output format

zedy-wj commented:
@jongio - The relevant changes in output have been updated. We are still a little confused about `AUTH_TYPE`. For details, please refer to my comments [here](https://github.com/jongio/chat-with-your-data-solution-accelerator/pull/2#discussion_r1449757392).
![image](https://github.com/jongio/chat-with-your-data-solution-accelerator/assets/81678720/c72d88d6-4f4f-45d6-bf70-11fb0511b3c9)

zedy-wj committed:
default to use rbac

jongio commented:
@zedy-wj - Are you ready for me to rereview this?

zedy-wj commented:
@jongio - Yes, it's ready for re-review.

jongio commented on README.md:
````diff
@@ -104,8 +104,46 @@ Out-of-the-box, you can upload the following file types:
+curl -fsSL https://aka.ms/install-azd.sh | bash
+```
+
+After logging in with the following command, you will be able to use azd cli to quickly provision and deploy the application.
````
```suggestion
After logging in with the following command, you will be able to use the `azd` cli to quickly provision and deploy the application.
```

jongio commented on app.py:
```diff
@@ -23,10 +25,20 @@
     return app.send_static_file(path)
 
+AUTH_TYPE = os.environ.get("AUTH_TYPE")
+
```
```suggestion
# Initialize Azure keys based on authentication type and environment settings. 
# When AUTH_TYPE is not 'rbac' and USE_KEY_VAULT environment variable is set, 
# Azure keys are securely fetched from Azure Key Vault using DefaultAzureCredential. 
# Otherwise, keys are obtained from environment variables, with fallbacks to None or 
# an empty string depending on the AUTH_TYPE.

```

jongio commented on app.py:
```diff
@@ -191,10 +203,10 @@ def stream_without_data(response):
-    openai.api_base = f"https://{AZURE_OPENAI_RESOURCE}.openai.azure.com/"
-    openai.api_version = "2023-03-15-preview"
-    openai.api_key = AZURE_OPENAI_KEY
+    if AUTH_TYPE == 'rbac':
+        openai_client = AzureOpenAI(azure_endpoint=f"https://{AZURE_OPENAI_RESOURCE}.openai.azure.com/",  api_version=AZURE_OPENAI_API_VERSION, azure_ad_token_provider=AZURE_TOKEN_PROVIDER)
+    else:
+        openai_client = AzureOpenAI(azure_endpoint=f"https://{AZURE_OPENAI_RESOURCE}.openai.azure.com/",  api_version=AZURE_OPENAI_API_VERSION, api_key=AZURE_OPENAI_KEY)
```
Please extract the azure endpoint string into a variable

jongio commented on backend/utilities/helpers/EnvHelper.py:
```diff
@@ -1,16 +1,28 @@
     def __init__(self, **kwargs) -> None:
         load_dotenv()
+
+        self.AUTH_TYPE =os.environ.get("AUTH_TYPE")
+        if not self.AUTH_TYPE == 'rbac' and os.environ.get("USE_KEY_VAULT"):
+            self.credential = DefaultAzureCredential()
+            self.secret_client = SecretClient(os.environ.get("AZURE_KEY_VAULT_ENDPOINT"), self.credential)
+            self.AZURE_SEARCH_KEY = self.secret_client.get_secret(os.environ.get("AZURE_SEARCH_KEY")).value
+            self.AZURE_OPENAI_KEY = self.secret_client.get_secret(os.environ.get("AZURE_OPENAI_KEY")).value
+        else:
+            self.AZURE_SEARCH_KEY = None if self.AUTH_TYPE == 'rbac' else os.environ.get("AZURE_SEARCH_KEY")
+            self.AZURE_OPENAI_KEY = "" if self.AUTH_TYPE == 'rbac' else os.environ.get("AZURE_OPENAI_KEY")
+            
```
I think this code is similar to code I've seen in this pr, can you see if there's a chance we can reuse this?

zedy-wj replied:
This code is used in two places, but they are `front-end` and `back-end` modules, which are independent of each other, so they cannot be reused.

jongio commented on backend/utilities/helpers/EnvHelper.py:
```diff
@@ -39,12 +50,13 @@ def __init__(self, **kwargs) -> None:
          # Set env for OpenAI SDK
-        self.OPENAI_API_TYPE = "azure"
+        self.OPENAI_API_TYPE = "azuread" if self.AUTH_TYPE == 'rbac' else "azure"
         self.OPENAI_API_BASE = f"https://{os.getenv('AZURE_OPENAI_RESOURCE')}.openai.azure.com/"
         self.OPENAI_API_KEY = self.AZURE_OPENAI_KEY
         self.OPENAI_API_VERSION = self.AZURE_OPENAI_API_VERSION
-        os.environ["OPENAI_API_TYPE"] = "azure"
+        os.environ["OPENAI_API_TYPE"] = "azuread" if self.AUTH_TYPE == 'rbac' else "azure"
```
These two conditions look like they're repeated, can you extract those into a variable?

jongio commented on infra/main.bicepparam:
```diff
@@ -0,0 +1,12 @@
+param principalId = readEnvironmentVariable('AZURE_PRINCIPAL_ID', 'principal_id')
+
+// Please make sure to set this value to false when using rbac with AUTH_TYPE
+param useKeyVault = bool(readEnvironmentVariable('USE_KEY_VAULT', 'false'))
```
default to true

jongio commented on app.py:
```diff
@@ -23,10 +25,20 @@
 def static_file(path):
     return app.send_static_file(path)
 
+AUTH_TYPE = os.environ.get("AUTH_TYPE")
```
Please default the AUTH_TYPE to keys and then add a section to the readme on how a user would switch from keys to rbac using azd env set AUTH_TYPE rbac.  Thanks

zedy-wj committed:
add some comments

jongio commented:
@zedy-wj @v-xuto There have been a significant amount of changes to the main repo: https://github.com/Azure-Samples/chat-with-your-data-solution-accelerator, since we started this PR.  They already implemented the RBAC / Keys solution in a different way.  (SOrry, I didn't know this was happening).

I pulled in the latest from upstream/main to jongio/main so you can see the difference.  

Can you please do the analysis to figure out how we can merge the two and what the differences are in implementation?

ChenxiJiang333 committed:
solve conflicts

jongio commented:
@zedy-wj - After the merge does the app now work?

zedy-wj committed:
add readme and speech service related code in main.bicep

zedy-wj commented:
@jongio - The merged code works well during local deployment. Since the directory structure of the project has also changed, we are solving the deployment structure problem in the `azure.yaml` file.

zedy-wj committed:
fix azd deploy problem

zedy-wj commented:
@jongio - We've updated the code and it's working fine so far, please re-review it, thanks!

jongio commented on azure.yaml:
```diff
@@ -3,25 +3,33 @@
+  prepackage:
+    windows:
+      shell: pwsh
+      run: robocopy -e ./code/utilities ./code/app/utilities; robocopy -e ./code/utilities ./code/admin/utilities; robocopy -e ./code/ ./code/admin/ requirements.txt /lev:1; robocopy -e ./code/utilities ./code/batch/utilities; robocopy -e ./code/ ./code/batch/ requirements.txt /lev:1;
```
Instead of doing this can you please move the code

jongio commented on code/app/app.py:
```diff
@@ -76,9 +64,22 @@ def get_config():
     AZURE_OPENAI_KEY = "" if AZURE_AUTH_TYPE == 'rbac' else os.environ.get("AZURE_OPENAI_KEY")
+    AZURE_SPEECH_KEY = None if AZURE_AUTH_TYPE == 'rbac' else os.environ.get("AZURE_SPEECH_SERVICE_KEY")
+
+@app.route('/api/config', methods=['GET'])
```
Can you move this up in the file so the diff looks better

zedy-wj committed:
move code folder structure

zedy-wj commented:
@jongio - We removed hooks and replaced them with structural changes in the `code` folder. Since the three services of `web`, `adminweb`, and `function` all require the `utilities` folder, we placed it in the innermost layer to ensure that other services are available during `azd deploy`. Currently both local deployment and azd deployment are working well, please re-review it, thanks!

jongio commented:
Thanks, I just updated main with latest from upstream, so there are merge conflicts now.

Can you give me an itemized list of the tests that you ran?  

Like
1. With RBAC
2. With Keys
3. With Keyvault = true
4. etc

zedy-wj committed:
fix conflict

zedy-wj commented:
@jongio - We resolved the conflict, and the update in `upstream` conflicts with our current code mainly in `rbac`. Since the upstream still uses an earlier version of `OpenAI`, this part retains the code in this PR.

> Can you give me an itemized list of the tests that you ran?

At present, we have the following three tests in the list, and different parameters need to be configured:
1. With RBAC
    - AZURE_AUTH_TYPE= rbac
    - USE_KEY_VAULT= false

2. With Keys and not use Key Vault
    - AZURE_AUTH_TYPE= keys
    - USE_KEY_VAULT= false

3. With Keys and use Key Vault
    - AZURE_AUTH_TYPE= keys
    - USE_KEY_VAULT= true

**Note:** The current default values ​​are: `AZURE_AUTH_TYPE=keys` and `USE_KEY_VAULT=true`.
